File: claude_exp/xv5/shared/crypto_utils.py
"""MasterCard M/Chip EMV operations. All functions are pure (no I/O)."""

import base64
import hashlib
import hmac

from Crypto.Cipher import DES, DES3

_MAC_FIELDS = (
    "amount_auth",
    "amount_other",
    "terminal_country",
    "terminal_verification_results",
    "currency_code",
    "transaction_date",
    "transaction_type",
    "unpredictable_number",
    "aip",
    "atc",
)


def derive_udk(imk_hex: str, pan: str, pan_seq: str) -> str:
    """EMV Option A UDK derivation."""
    imk = bytes.fromhex(imk_hex)
    pan_psn = (pan + pan_seq)[-16:].rjust(16, "0")
    z = bytes.fromhex(pan_psn)
    cipher = DES3.new(imk, DES3.MODE_ECB)
    udk_a = cipher.encrypt(z)
    z_complement = bytes(b ^ 0xFF for b in z)
    udk_b = cipher.encrypt(z_complement)
    return (udk_a + udk_b).hex().upper()


def derive_session_key(udk_hex: str, atc_hex: str) -> str:
    """ATC-based session key (EMV Common Session Key Derivation, Option A)."""
    udk = bytes.fromhex(udk_hex)
    cipher = DES3.new(udk, DES3.MODE_ECB)
    atc = bytes.fromhex(atc_hex.zfill(4))[:2]
    r_left = atc + b"\xf0" + b"\x00" * 5
    r_right = atc + b"\x0f" + b"\x00" * 5
    sk_left = cipher.encrypt(r_left)
    sk_right = cipher.encrypt(r_right)
    return (sk_left + sk_right).hex().upper()


def _pad_iso9797_2(data: bytes) -> bytes:
    padded = data + b"\x80"
    while len(padded) % 8 != 0:
        padded += b"\x00"
    return padded


def _retail_mac(key16: bytes, data: bytes) -> bytes:
    """ISO/IEC 9797-1 MAC Algorithm 3 (Retail MAC)."""
    k1, k2 = key16[:8], key16[8:16]
    des_k1 = DES.new(k1, DES.MODE_ECB)
    des_k2 = DES.new(k2, DES.MODE_ECB)
    h = b"\x00" * 8
    for i in range(0, len(data), 8):
        block = data[i : i + 8]
        x = bytes(a ^ b for a, b in zip(block, h))
        h = des_k1.encrypt(x)
    return des_k1.encrypt(des_k2.decrypt(h))


def _build_mac_input(f55: dict) -> bytes:
    data = b""
    for key in _MAC_FIELDS:
        data += bytes.fromhex(f55[key])
    return data


def verify_arqc(pan: str, pan_seq: str, imk_hex: str, f55: dict) -> bool:
    """Retail MAC ARQC check."""
    udk = derive_udk(imk_hex, pan, pan_seq)
    sk = derive_session_key(udk, f55["atc"])
    data = _pad_iso9797_2(_build_mac_input(f55))
    mac = _retail_mac(bytes.fromhex(sk), data)
    return mac.hex().upper() == f55.get("cryptogram", "").upper()


def calculate_arpc_method1(arqc_hex: str, arc_hex: str, sk_hex: str) -> bytes:
    """ARPC Method 1: encrypt(ARQC XOR (ARC padded to 8 bytes)) with the session key."""
    arqc = bytearray(bytes.fromhex(arqc_hex))
    arc = bytes.fromhex(arc_hex.zfill(4))
    for i in range(min(len(arc), len(arqc))):
        arqc[i] ^= arc[i]
    cipher = DES3.new(bytes.fromhex(sk_hex), DES3.MODE_ECB)
    return cipher.encrypt(bytes(arqc))


def _pan_block(pan: str) -> bytes:
    digits = pan[:-1][-12:].rjust(12, "0")
    return bytes.fromhex("0000" + digits)


def encode_pin_block_format0(pin: str, pan: str) -> bytes:
    """Build cleartext ISO 9564-1 Format-0 PIN block (tests)."""
    control = "0"
    length_nibble = format(len(pin), "X")
    pin_field = (control + length_nibble + pin).ljust(16, "F")
    pin_block = bytes.fromhex(pin_field)
    pan_block = _pan_block(pan)
    return bytes(a ^ b for a, b in zip(pin_block, pan_block))


def encrypt_pin_block(plain: bytes, pek_hex: str) -> bytes:
    """3DES encrypt PIN block."""
    cipher = DES3.new(bytes.fromhex(pek_hex), DES3.MODE_ECB)
    return cipher.encrypt(plain)


def verify_pin(pan: str, f52_b64: str, pek_hex: str, reference_pin: str) -> bool:
    """ISO 9564-1 Format-0 PIN block verification."""
    try:
        encrypted = base64.b64decode(f52_b64)
    except Exception:
        return False
    cipher = DES3.new(bytes.fromhex(pek_hex), DES3.MODE_ECB)
    clear_block = cipher.decrypt(encrypted)
    pan_block = _pan_block(pan)
    pin_block = bytes(a ^ b for a, b in zip(clear_block, pan_block))
    hexstr = pin_block.hex().upper()
    pin_len = int(hexstr[1], 16)
    pin_digits = hexstr[2 : 2 + pin_len]
    return pin_digits == reference_pin


def _cvv_data_blocks(pan: str, expiry_mmyy: str, service_code: str) -> tuple:
    mm, yy = expiry_mmyy[:2], expiry_mmyy[2:]
    expiry_yymm = yy + mm
    data = (pan + expiry_yymm + service_code).ljust(32, "0")[:32]
    return bytes.fromhex(data[:16]), bytes.fromhex(data[16:32])


def _cvv_core(pan: str, expiry_mmyy: str, cvk_hex: str, service_code: str = "000") -> str:
    cvk = bytes.fromhex(cvk_hex)
    cvk_a, cvk_b = cvk[:8], cvk[8:16]
    block_a, block_b = _cvv_data_blocks(pan, expiry_mmyy, service_code)
    des_a = DES.new(cvk_a, DES.MODE_ECB)
    des_b = DES.new(cvk_b, DES.MODE_ECB)
    r1 = des_a.encrypt(block_a)
    r2 = bytes(a ^ b for a, b in zip(r1, block_b))
    r3 = des_a.encrypt(r2)
    r4 = des_b.decrypt(r3)
    r5 = des_a.encrypt(r4)
    hexdigits = r5.hex()
    decimals = [c for c in hexdigits if c.isdigit()]
    if len(decimals) < 3:
        decimals += [str((int(c, 16) - 10) % 10) for c in hexdigits if c.upper() in "ABCDEF"]
    return "".join(decimals[:3])


def compute_cvv2(pan: str, expiry_mmyy: str, cvk_hex: str) -> str:
    """Compute CVV2 (tests)."""
    return _cvv_core(pan, expiry_mmyy, cvk_hex)


def verify_cvv2(pan: str, expiry_mmyy: str, cvv2: str, cvk_hex: str) -> bool:
    """MasterCard CVV2 verification."""
    return compute_cvv2(pan, expiry_mmyy, cvk_hex) == cvv2


def compute_aav(f47_data: dict, aav_key_hex: str, pan: str) -> str:
    """Compute AAV (tests)."""
    key = bytes.fromhex(aav_key_hex)
    message = (pan + f47_data.get("f14", "") + f47_data.get("message_type", "")).encode("ascii")
    mac = hmac.new(key, message, hashlib.sha1).digest()
    return base64.b64encode(mac).decode("ascii")


def verify_aav(f47_data: dict, aav_key_hex: str, pan: str) -> bool:
    """HMAC-SHA1 AAV verification."""
    return f47_data.get("aav", "") == compute_aav(f47_data, aav_key_hex, pan)

File: claude_exp/xv6java/src/main/java/com/xv6/shared/CryptoUtils.java
package com.xv6.shared;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.HexFormat;
import java.util.Map;

/**
 * MasterCard M/Chip EMV operations. All methods are pure (no I/O). Port of xv5's
 * shared/crypto_utils.py, using standard JCE ({@code DESede/ECB/NoPadding}, {@code
 * DES/ECB/NoPadding}, {@code HmacSHA1}) - all present in the default SunJCE provider on JDK >=
 * 8u162, no BouncyCastle needed.
 */
public final class CryptoUtils {

    private static final HexFormat HEX = HexFormat.of();

    private CryptoUtils() {
    }

    private static Cipher des3Ecb() {
        try {
            return Cipher.getInstance("DESede/ECB/NoPadding");
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private static Cipher desEcb() {
        try {
            return Cipher.getInstance("DES/ECB/NoPadding");
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private static byte[] des3Encrypt(byte[] key16or24, byte[] data) {
        try {
            Cipher c = des3Ecb();
            c.init(Cipher.ENCRYPT_MODE, desedeKey(key16or24));
            return c.doFinal(data);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private static byte[] des3Decrypt(byte[] key16or24, byte[] data) {
        try {
            Cipher c = des3Ecb();
            c.init(Cipher.DECRYPT_MODE, desedeKey(key16or24));
            return c.doFinal(data);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private static SecretKeySpec desedeKey(byte[] key) throws InvalidKeyException {
        // Unlike pycryptodome (which accepts a 16-byte key directly for two-key triple-DES),
        // the SunJCE DESede cipher requires exactly 24 bytes - a 16-byte key must be expanded to
        // K1||K2||K1 first (verified: a bare 16-byte SecretKeySpec throws "Wrong key size").
        if (key.length == 16) {
            byte[] expanded = new byte[24];
            System.arraycopy(key, 0, expanded, 0, 16);
            System.arraycopy(key, 0, expanded, 16, 8);
            key = expanded;
        }
        return new SecretKeySpec(key, "DESede");
    }

    private static byte[] desEncrypt(byte[] key8, byte[] data) {
        try {
            Cipher c = desEcb();
            c.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key8, "DES"));
            return c.doFinal(data);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private static byte[] desDecrypt(byte[] key8, byte[] data) {
        try {
            Cipher c = desEcb();
            c.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key8, "DES"));
            return c.doFinal(data);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private static byte[] xor(byte[] a, byte[] b) {
        int n = Math.min(a.length, b.length);
        byte[] out = new byte[n];
        for (int i = 0; i < n; i++) {
            out[i] = (byte) (a[i] ^ b[i]);
        }
        return out;
    }

    private static byte[] hex(String s) {
        return HEX.parseHex(s);
    }

    private static String hexUpper(byte[] b) {
        return HEX.formatHex(b).toUpperCase();
    }

    /** EMV Option A UDK derivation. */
    public static String deriveUdk(String imkHex, String pan, String panSeq) {
        byte[] imk = hex(imkHex);
        String panPsn = rightJustify16((pan + panSeq));
        byte[] z = hex(panPsn);
        byte[] udkA = des3Encrypt(imk, z);
        byte[] zComplement = new byte[z.length];
        for (int i = 0; i < z.length; i++) {
            zComplement[i] = (byte) (~z[i] & 0xFF);
        }
        byte[] udkB = des3Encrypt(imk, zComplement);
        byte[] udk = concat(udkA, udkB);
        return hexUpper(udk);
    }

    private static String rightJustify16(String s) {
        String tail = s.length() > 16 ? s.substring(s.length() - 16) : s;
        StringBuilder sb = new StringBuilder();
        for (int i = tail.length(); i < 16; i++) {
            sb.append('0');
        }
        return sb.append(tail).toString();
    }

    /** ATC-based session key (EMV Common Session Key Derivation, Option A). */
    public static String deriveSessionKey(String udkHex, String atcHex) {
        byte[] udk = hex(udkHex);
        // Python: bytes.fromhex(atc_hex.zfill(4))[:2] - first two bytes of the zero-padded,
        // hex-decoded ATC (not the last two).
        byte[] atcFull = hex(zfill(atcHex, 4));
        byte[] atc = {atcFull[0], atcFull.length > 1 ? atcFull[1] : 0};

        byte[] rLeft = concat(atc, new byte[]{(byte) 0xF0}, new byte[5]);
        byte[] rRight = concat(atc, new byte[]{(byte) 0x0F}, new byte[5]);
        byte[] skLeft = des3Encrypt(udk, rLeft);
        byte[] skRight = des3Encrypt(udk, rRight);
        return hexUpper(concat(skLeft, skRight));
    }

    private static byte[] concat(byte[]... parts) {
        int total = 0;
        for (byte[] p : parts) {
            total += p.length;
        }
        byte[] out = new byte[total];
        int off = 0;
        for (byte[] p : parts) {
            System.arraycopy(p, 0, out, off, p.length);
            off += p.length;
        }
        return out;
    }

    // Package-private (not private) so CryptoUtilsTest, in the same package, can build a
    // self-consistent ARQC the same way xv5's white-box test_crypto_utils.py does via direct
    // access to _pad_iso9797_2/_build_mac_input/_retail_mac.
    static byte[] padIso9797_2(byte[] data) {
        int padded = data.length + 1;
        while (padded % 8 != 0) {
            padded++;
        }
        byte[] out = new byte[padded];
        System.arraycopy(data, 0, out, 0, data.length);
        out[data.length] = (byte) 0x80;
        return out;
    }

    /** ISO/IEC 9797-1 MAC Algorithm 3 (Retail MAC). */
    static byte[] retailMac(byte[] key16, byte[] data) {
        byte[] k1 = new byte[8];
        byte[] k2 = new byte[8];
        System.arraycopy(key16, 0, k1, 0, 8);
        System.arraycopy(key16, 8, k2, 0, 8);

        byte[] h = new byte[8];
        for (int i = 0; i < data.length; i += 8) {
            byte[] block = new byte[8];
            System.arraycopy(data, i, block, 0, 8);
            byte[] x = xor(block, h);
            h = desEncrypt(k1, x);
        }
        return desEncrypt(k1, desDecrypt(k2, h));
    }

    private static final String[] MAC_FIELDS = {
            "amount_auth", "amount_other", "terminal_country", "terminal_verification_results",
            "currency_code", "transaction_date", "transaction_type", "unpredictable_number", "aip", "atc",
    };

    static byte[] buildMacInput(Map<String, Object> f55) {
        int total = 0;
        byte[][] parts = new byte[MAC_FIELDS.length][];
        for (int i = 0; i < MAC_FIELDS.length; i++) {
            parts[i] = hex(String.valueOf(f55.get(MAC_FIELDS[i])));
            total += parts[i].length;
        }
        byte[] out = new byte[total];
        int off = 0;
        for (byte[] p : parts) {
            System.arraycopy(p, 0, out, off, p.length);
            off += p.length;
        }
        return out;
    }

    /** Retail MAC ARQC check. */
    public static boolean verifyArqc(String pan, String panSeq, String imkHex, Map<String, Object> f55) {
        String udk = deriveUdk(imkHex, pan, panSeq);
        String sk = deriveSessionKey(udk, String.valueOf(f55.getOrDefault("atc", "0000")));
        byte[] data = padIso9797_2(buildMacInput(f55));
        byte[] mac = retailMac(hex(sk), data);
        Object cryptogram = f55.get("cryptogram");
        String expected = cryptogram == null ? "" : String.valueOf(cryptogram);
        return hexUpper(mac).equals(expected.toUpperCase());
    }

    /** ARPC Method 1: encrypt(ARQC XOR (ARC padded to 8 bytes)) with the session key. */
    public static byte[] calculateArpcMethod1(String arqcHex, String arcHex, String skHex) {
        byte[] arqc = hex(arqcHex).clone();
        byte[] arc = hex(zfill(arcHex, 4));
        int n = Math.min(arc.length, arqc.length);
        for (int i = 0; i < n; i++) {
            arqc[i] ^= arc[i];
        }
        return des3Encrypt(hex(skHex), arqc);
    }

    private static byte[] panBlock(String pan) {
        String withoutCheckDigit = pan.substring(0, pan.length() - 1);
        String tail = withoutCheckDigit.length() > 12
                ? withoutCheckDigit.substring(withoutCheckDigit.length() - 12)
                : withoutCheckDigit;
        String digits = zfill(tail, 12);
        return hex("0000" + digits);
    }

    /** Build cleartext ISO 9564-1 Format-0 PIN block (tests). */
    public static byte[] encodePinBlockFormat0(String pin, String pan) {
        String control = "0";
        String lengthNibble = Integer.toHexString(pin.length()).toUpperCase();
        StringBuilder pinField = new StringBuilder(control).append(lengthNibble).append(pin);
        while (pinField.length() < 16) {
            pinField.append('F');
        }
        byte[] pinBlock = hex(pinField.toString());
        byte[] panBlock = panBlock(pan);
        return xor(pinBlock, panBlock);
    }

    /** 3DES encrypt PIN block. */
    public static byte[] encryptPinBlock(byte[] plain, String pekHex) {
        return des3Encrypt(hex(pekHex), plain);
    }

    /** ISO 9564-1 Format-0 PIN block verification. */
    public static boolean verifyPin(String pan, String f52Base64, String pekHex, String referencePin) {
        byte[] encrypted;
        try {
            encrypted = Base64.getDecoder().decode(f52Base64);
        } catch (Exception e) {
            return false;
        }
        byte[] clearBlock = des3Decrypt(hex(pekHex), encrypted);
        byte[] panBlock = panBlock(pan);
        byte[] pinBlock = xor(clearBlock, panBlock);
        String hexStr = hexUpper(pinBlock);
        int pinLen = Character.digit(hexStr.charAt(1), 16);
        if (pinLen < 0 || 2 + pinLen > hexStr.length()) {
            return false;
        }
        String pinDigits = hexStr.substring(2, 2 + pinLen);
        return pinDigits.equals(referencePin);
    }

    private static byte[][] cvvDataBlocks(String pan, String expiryMmyy, String serviceCode) {
        String mm = expiryMmyy.substring(0, 2);
        String yy = expiryMmyy.substring(2);
        String expiryYymm = yy + mm;
        String data = (pan + expiryYymm + serviceCode);
        data = data.length() > 32 ? data.substring(0, 32) : rightPad(data, 32, '0');
        return new byte[][]{hex(data.substring(0, 16)), hex(data.substring(16, 32))};
    }

    private static String rightPad(String s, int width, char c) {
        StringBuilder sb = new StringBuilder(s);
        while (sb.length() < width) {
            sb.append(c);
        }
        return sb.toString();
    }

    private static String cvvCore(String pan, String expiryMmyy, String cvkHex, String serviceCode) {
        byte[] cvk = hex(cvkHex);
        byte[] cvkA = new byte[8];
        byte[] cvkB = new byte[8];
        System.arraycopy(cvk, 0, cvkA, 0, 8);
        System.arraycopy(cvk, 8, cvkB, 0, 8);

        byte[][] blocks = cvvDataBlocks(pan, expiryMmyy, serviceCode);
        byte[] r1 = desEncrypt(cvkA, blocks[0]);
        byte[] r2 = xor(r1, blocks[1]);
        byte[] r3 = desEncrypt(cvkA, r2);
        byte[] r4 = desDecrypt(cvkB, r3);
        byte[] r5 = desEncrypt(cvkA, r4);

        String hexDigits = HEX.formatHex(r5);
        StringBuilder decimals = new StringBuilder();
        for (char c : hexDigits.toCharArray()) {
            if (Character.isDigit(c)) {
                decimals.append(c);
            }
        }
        if (decimals.length() < 3) {
            for (char c : hexDigits.toCharArray()) {
                if (Character.isLetter(c)) {
                    int v = (Character.digit(c, 16) - 10) % 10;
                    decimals.append(v);
                }
            }
        }
        return decimals.substring(0, Math.min(3, decimals.length()));
    }

    /** Compute CVV2 (tests). */
    public static String computeCvv2(String pan, String expiryMmyy, String cvkHex) {
        return cvvCore(pan, expiryMmyy, cvkHex, "000");
    }

    /** MasterCard CVV2 verification. */
    public static boolean verifyCvv2(String pan, String expiryMmyy, String cvv2, String cvkHex) {
        return computeCvv2(pan, expiryMmyy, cvkHex).equals(cvv2);
    }

    /** Compute AAV (tests). */
    public static String computeAav(Map<String, Object> f47Data, String aavKeyHex, String pan) {
        try {
            byte[] key = hex(aavKeyHex);
            String f14 = String.valueOf(f47Data.getOrDefault("f14", ""));
            String messageType = String.valueOf(f47Data.getOrDefault("message_type", ""));
            byte[] message = (pan + f14 + messageType).getBytes(StandardCharsets.US_ASCII);

            Mac mac = Mac.getInstance("HmacSHA1");
            mac.init(new SecretKeySpec(key, "HmacSHA1"));
            byte[] digest = mac.doFinal(message);
            return Base64.getEncoder().encodeToString(digest);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new IllegalStateException(e);
        }
    }

    /** HMAC-SHA1 AAV verification. */
    public static boolean verifyAav(Map<String, Object> f47Data, String aavKeyHex, String pan) {
        String aav = String.valueOf(f47Data.getOrDefault("aav", ""));
        return aav.equals(computeAav(f47Data, aavKeyHex, pan));
    }

    private static String zfill(String s, int width) {
        StringBuilder sb = new StringBuilder();
        for (int i = s.length(); i < width; i++) {
            sb.append('0');
        }
        return sb.append(s).toString();
    }
}
